newmail_notifier: Include sender and subject in desktop notification

Daniel15 opened the pull request:
Displays the sender and subject in desktop notifications:
![image](https://github.com/roundcube/roundcubemail/assets/91933/000ba235-01da-47a4-8a83-4e906ef2c2af)

References #8360 #7826 #8587

Daniel15 committed:
newmail_notifier: Include sender and subject in desktop notification

Daniel15 commented on plugins/newmail_notifier/newmail_notifier.js:
```diff
@@ -37,7 +37,19 @@ function newmail_notifier_run(prop) {
     }
     if (prop.desktop) {
-        newmail_notifier_desktop(rcmail.get_label('body', 'newmail_notifier'));
+        var title = prop.from || rcmail.get_label('title', 'newmail_notifier');
```
The JavaScript in Roundcube seems to not be using newer features, so I'm avoiding them in my changes (`var` instead of `const`/`let`  here, and string concatenation instead of template literals for the strings on line 48 and 50)

pabzm replied:
That's thoughtful, thanks!

pabzm commented on plugins/newmail_notifier/newmail_notifier.js:
```diff
@@ -37,7 +37,19 @@ function newmail_notifier_run(prop) {
+        if (otherCount >= 1) {
+            title += prop.from
+                // Have sender name, e.g. "Daniel L +3"
+                ? ' +' +otherCount
+                // No sender name, e.g. "New Mail! (3)
+                : '('+otherCount + ')';
```
Please add spaces around the plus-signs. Or maybe you could rework these title-related lines a little, including line 40, to be a little less terse but more readable?

pabzm reviewed (changes requested):
Thank you for the contribution, I like it!

From my side there's only the minor nitpick, the rest looks good!

alecpl commented on plugins/newmail_notifier/newmail_notifier.php:
```diff
@@ -203,12 +203,31 @@ public function notify($args)
+            if ($latest_uid !== null) {
+                $headers = $storage->fetch_headers($mbox, [$latest_uid]);
+                // fetch_headers returns an array, but we only care about the first one.
+                $headers = array_shift($headers);
```
Not a big deal, but here you could use `array_first()`.

alecpl commented on plugins/newmail_notifier/newmail_notifier.php:
```diff
@@ -203,12 +203,31 @@ public function notify($args)
+                $headers = array_shift($headers);
+                if ($headers !== null) {
+                    $from = $headers->from;
+                    $subject = $headers->subject;
```
I wonder what sanity checks we should apply here. We probably should not use too long subjects. Also we should parse the From header content. We definitely should decode the content, not pass as-is.

See program/actions/mail/show.php for hints on what we normally do with subject and from header.

Daniel15 replied:
Thanks, I'll take a look. I wasn't sure whether to limit the subject length here or on the client side.

Daniel15 replied:
> We probably should not use too long subjects.

I tried with a long subject, and KDE makes them scrollable:

![image](https://github.com/roundcube/roundcubemail/assets/91933/4a876856-8cf3-4b25-95a8-c577efbd7c00)

I haven't tested with Windows yet, but I think it truncates them.

> We definitely should decode the content, not pass as-is. See program/actions/mail/show.php for hints on what we normally do with subject and from header.

Thanks for the pointer. I thought the headers would already be decoded here.

alecpl commented on plugins/newmail_notifier/newmail_notifier.js:
```diff
@@ -37,7 +37,19 @@ function newmail_notifier_run(prop) {
+        var otherCount = prop.count - 1;
+        if (otherCount >= 1) {
+            title += prop.from
+                // Have sender name, e.g. "Daniel L +3"
```
I don't like this whole thing. Neither "Daniel L +3" looks good nor "New Mail! (3)".

How about we show the sender/subject only when there's a single new mail message? And "(n) New Mail!" in other cases?

Daniel15 commented:
I haven't had free time to revise this yet, but I'll get back to it soon.

Daniel15 committed:
Merge remote branch “upstream/master”

Daniel15 committed:
Code review comments

 - Show "(2) New Mail!" when there's more than one email
 - Decode from and subject headers

Signed-off-by: Daniel Lo Nigro <git@d.sb>

Daniel15 commented:
Updated to properly decode the from and subject headers. I updated it to just show the sender name rather than their email address too, to make the notification a bit more concise.

Tested with a multi-byte (Chinese) subject:
![Screenshot_20240707_175454](https://github.com/roundcube/roundcubemail/assets/91933/c9a1b768-e14d-416f-bf26-186e4db5b242)

Tested with a Chinese sender name:
![Screenshot_20240707_175550](https://github.com/roundcube/roundcubemail/assets/91933/468d7899-5d37-4f4b-ac29-9ce660bb5787)

alecpl commented on plugins/newmail_notifier/newmail_notifier.php:
```diff
@@ -203,12 +203,36 @@ public function notify($args)
+                    $from_details = array_first(
+                        rcube_mime::decode_address_list($headers->from, 1, true, $headers->charset)
+                    );
+                    $from = $from_details['name'];
```
Name can be empty. In such case let's fallback to the email address. Also, in some corner cases `From` header might not exist at all.

pabzm commented:
@Daniel15 Are you still interested in this? If so, please have a look at Alec's latest comments from July. If not, please close.

pabzm commented:
@Daniel15 I'm interested in bringing this forward and would offer to take over in about three weeks, if you can't find time for it.
